src/thunderbots/software/network_input/filter/ball_filter.cpp

```diff
@@ -1,35 +1,214 @@
+                // If we determine the data to be noise, remove an entry from the buffer.
+                // This way if we have messed up and now the ball is too far away for the
+                // buffer to track, the buffer will rapidly shrink and start tracking the
+                // ball at its new location once the buffer is empty.
+                // We sort the vector in decreasing order first so that we can always ensure any elements that are
+                // ejected from the end of the buffer are the oldest data
+                std::sort(ball_detection_buffer.rbegin(), ball_detection_buffer.rend());
+                ball_detection_buffer.pop_back();
+            }else {
+                // We sort the vector in decreasing order first so that we can always ensure any elements that are
+                // ejected from the end of the buffer are the oldest data
+                std::sort(ball_detection_buffer.rbegin(), ball_detection_buffer.rend());
+                ball_detection_buffer.push_front(detection);
+            }
+        }else {
+            // If there is no data in the buffer, we always add the new data
+            // We sort the vector in decreasing order first so that we can always ensure any elements that are
+            // ejected from the end of the buffer are the oldest data
+            std::sort(ball_detection_buffer.rbegin(), ball_detection_buffer.rend());
+            ball_detection_buffer.push_front(detection);
+        }
+    }
+}
+
+std::vector<Vector> BallFilter::calculateBallVelocities(
+        boost::circular_buffer<SSLBallDetection> ball_detections) {
+    // Sort the detections in decreasing order before processing. This places the most recent detections (with the
+    // largest timestamp) at the front of the buffer, and the oldest detections (smallest timestamp) at the end of
+    // the buffer
+    std::sort(ball_detections.rbegin(), ball_detections.rend());
+
+    std::vector<Vector> ball_velocities;
+    for(int i = 0; i < ball_detections.size() - 1; i++) {
+        SSLBallDetection current_detection = ball_detections.at(i);
+        SSLBallDetection previous_detection = ball_detections.at(i + 1);
+
+        Duration time_diff = current_detection.timestamp - previous_detection.timestamp;
+        // Avoid division by 0. If we have adjacent detections with the same timestamp the velocity cannot be
+        // calculated
+        if(time_diff.getSeconds() == 0){
+            continue;
+        }
+
+        Vector velocity_vector = current_detection.position - previous_detection.position;
+        double velocity_magnitude = velocity_vector.len() / time_diff.getSeconds();
+        Vector velocity = velocity_vector.norm(velocity_magnitude);
+
+        ball_velocities.emplace_back(velocity);
+    }
+
+    return ball_velocities;
+}
+
+Ball
+BallFilter::getLinearRegressionPositionAndVelocity(boost::circular_buffer<SSLBallDetection> ball_detections) {
```

Comment:
This is internal to the filter and was originally named this way because there were two filter options (averaging and linear regression). Do you still want to change this?

Reply:
Since there's only one, yeah, I think still changing this would be good.

Reply:
fixed